src/convert_hl7v2_fhir/controllers/hl7/hl7_message_controller.py

```diff
@@ -0,0 +1,262 @@
+                "address": {
+                    "line": location_name,
+                    "city": self.metadata["location"]["address"]["city"],
+                    "postalCode": self.metadata["location"]["address"]["postalCode"],
+                },
+            },
+        }
+
+    def _create_organization(self):
+        return {
+            "fullUrl": f"urn:uuid:{self.organization_uuid}",
+            "resource": {
+                "resourceType": "Organization",
+                "id": str(self.organization_uuid),
+                "meta": {
+                    "profile": [
+                        "https://fhir.hl7.org.uk/StructureDefinition/UKCore-Organization"
+                    ]
+                },
+                "identifier": [
+                    {
+                        "system": "https://fhir.nhs.uk/Id/ods-organization-code",
+                        "value": self.metadata["organization"]["identifier"]["value"],
+                    }
+                ],
+                "name": self.metadata["organization"]["name"],
+            },
+        }
+
+    def _create_encounter(self):
+        resource_class = to_fhir_encounter_class(self._extract_field("PV1", 0, 2))
+        resource_period_start = to_fhir_datetime(self._extract_field("PV1", 0, 44))
+        admission_method_coding = to_fhir_admission_method(
+            self._extract_field("PV1", 0, 4)
+        )
+        return {
+            "fullUrl": f"urn:uuid:{self.encounter_uuid}",
+            "resource": {
+                "resourceType": "Encounter",
+                "id": str(self.encounter_uuid),
+                "meta": {
+                    "profile": [
+                        "https://fhir.simplifier.net/Hospital-Activity-Notification-Service/StructureDefinition/ActivityNotification-UKCore-Encounter"
+                    ]
+                },
+                "status": "in-progress",
+                "subject": {"reference": f"urn:uuid:{self.patient_uuid}"},
+                "class": resource_class,
+                "period": {"start": resource_period_start},
+                "location": [
+                    {
+                        "status": "active",
+                        "location": {"reference": f"urn:uuid:{self.location_uuid}"},
+                    }
+                ],
+                "extension": [
+                    {
+                        "url": "https://fhir.hl7.org.uk/StructureDefinition/Extension-UKCore-AdmissionMethod",
+                        "valueCodeableConcept": {"coding": [admission_method_coding]},
+                    }
+                ],
+            },
+        }
+
+    def _extract_nhs_number(self):
+        pid_segment = self._extract_field("PID")
+        potential_nhs_numbers = set(re.findall(r"\d{10}", pid_segment))
+        if "NHSNMBR" not in pid_segment or not potential_nhs_numbers:
+            raise MissingNHSNumberError
+
+        try:
+            return next(

```

Comment:
This feels fragile. We're identifying whether `NHSNMBR` appears anywhere in the `pid_segment` then scanning for any 10-digit sequence that might be anywhere else in the `pid_segment` - there's no reason for the two to be coming from the same place.  In our example below if the `SIMULATOR MRN` value happened to pass the NHS number checksum calculation, that's what we'd get returned.

Reply:
Yep this is definitely not what we want - @michal-kras can you look at the current `get_nhs_number()` method and extract the logic from there?

It's in [the current version of the code on the repo](https://github.com/nhsx/il-hans-nems-subscriber/blob/main/src/convert_hl7v2_fhir/controllers/hl7utils.py)